raft/node.go

```diff
@@ -49,43 +49,90 @@ func (a *SoftState) equal(b *SoftState) bool {
 // Ready encapsulates the entries and messages that are ready to read,
 // be saved to stable storage, committed or sent to other peers.
 // All fields in Ready are read-only.
+//
+// Handling the Ready can be done with or without performance optimizations.
+// Applications that don't need high throughput should eschew performance
+// optimizations for simplicity. These applications will use the following
+// sequence of steps:
+//
+// 1 apply snapshot (if any) and sync
+// 2 atomically append Entries (if any) and update HardState (if any).
+// 3. sync the Entries and HardState if MustSync is set.
+// 4. send messages
+// 5. apply CommittedEntries
+//
+// If for some reason it is impossible to append the Entries atomically with
+// updating the HardState, the entries (which are always synced) must be
+// appended and synced before the HardState is updated. This avoids the
+// situation in which the HardState references a Commit index which is not
+// durably appended yet. This does not penalize performance because updates to
+// the HardState must only be synced when MustSync is set, and those occur only
+// during elections and membership changes.
+//
+// For performance, applications may also split the messages to be sent in
+// MsgApp and a remainder, and introduce a step 1½ in the above list:
+//
+// 1.  unchanged
+// 1½. send all MsgApp
+// 2.  unchanged
+// 3.  unchanged
+// 4.  send remaining messages
+// 5.  unchanged
+//
+// This reduces replication latency because the MsgApp are no longer delayed by
+// the disk sync in step 3. This optimization is safe because a corresponding
+// MsgAppResp can (at the earliest) be processed in the next Ready cycle, at
+// which point the previous Ready is fully synced. Should the leader (which is
+// the only peer sending MsgApp) have crashed after sending the MsgApp but
+// before persisting its Entries, it won't be the leader for its old term any
+// more and thus will not handle the MsgAppResp. Additionally, the fact that a
+// commit index that is higher than that ultimately persisted on the crashed
+// leader is now known to a follower (via the MsgApp) is benign. And, for one
```

Comment:
> "the fact ... is benign"

Is that what you meant to say?

Reply:
Heh, not quite. Will fix.

Going to ping when it's ready, won't be today